File: modeler/src/main/java/net/sayaya/document/modeler/model/ModelHandler.java
package net.sayaya.document.modeler.model;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import net.sayaya.document.data.Model;
import net.sayaya.document.data.MessageModel;
import org.springframework.context.annotation.Bean;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.publisher.Sinks;

import java.util.function.Consumer;
import java.util.function.Supplier;

@Service
public class ModelHandler {
	private final ModelRepository repo;
	private final ObjectMapper OM;
	private final Sinks.Many<MessageModel> publisher = Sinks.many().unicast().onBackpressureBuffer();
	private final Sinks.Many<MessageModel> subscriber = Sinks.many().multicast().directBestEffort();
	public ModelHandler(ModelRepository repo, ObjectMapper om) {
		this.repo = repo;
		OM = om;
	}
	public Flux<Model> list() {
		return repo.findAll().map(ModelToDTO::map);
	}
	public Mono<MessageModel> create(String name) {
		net.sayaya.document.modeler.model.Model model = new net.sayaya.document.modeler.model.Model();
		model.setName(name);
		return repo.save(model)
				.map(ModelToDTO::map)
				.map(data->new MessageModel(MessageModel.MessageType.CREATE, data))
				.doOnSuccess(publisher::tryEmitNext);
	}
	public Mono<MessageModel> remove(String name) {
		return repo.deleteById(name)
				.then(Mono.just(new Model().name(name).cntDocuments(0)))
				.map(data->new MessageModel(MessageModel.MessageType.DELETE, data))
				.doOnSuccess(publisher::tryEmitNext);
	}
	public Flux<MessageModel> subscribe() {
		return subscriber.asFlux();
	}
	private String map(MessageModel dto) {
		try {
			return OM.writeValueAsString(dto);
		} catch(JsonProcessingException e) {
			throw new RuntimeException(e.getMessage(), e);
		}
	}
	private MessageModel map(String json) {
		try {
			return OM.readValue(json, MessageModel.class);
		} catch(JsonProcessingException e) {
			throw new RuntimeException(e.getMessage(), e);
		}
	}
	@Bean("publish-model")
	public Supplier<Flux<String>> publishModel() {
		return ()-> publisher.asFlux().map(this::map);
	}
	@Bean("broadcast-model")
	public Consumer<String> broadcastModel() {
		return c->subscriber.tryEmitNext(map(c));
	}
}

File: modeler/src/main/kotlin/net/sayaya/document/modeler/model/ModelHandler.kt
package net.sayaya.document.modeler.model

import com.fasterxml.jackson.databind.ObjectMapper
import net.sayaya.document.data.MessageModel
import org.mapstruct.factory.Mappers
import org.springframework.context.annotation.Bean
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.core.publisher.Sinks
import java.util.function.Consumer
import java.util.function.Supplier

@Service
class ModelHandler(
    private val repo: ModelRepository,
    private val om: ObjectMapper,
    private val mapper: ModelToDTO
) {
    private val publisher = Sinks.many().unicast().onBackpressureBuffer<MessageModel>()
    private val subscriber = Sinks.many().multicast().directBestEffort<MessageModel>()
    fun list(): Flux<net.sayaya.document.data.Model> {
        return repo.findAll().map(mapper::toModel)
    }
    fun create(name: String): Mono<MessageModel> {
        val model = Model(name)
        return repo.save(model).map(mapper::toModel)
            .map { data -> MessageModel(MessageModel.MessageType.CREATE, data) }
            .doOnSuccess(publisher::tryEmitNext)
    }
    fun remove(name: String): Mono<MessageModel> {
        return repo.deleteById(name)
            .then(Mono.just(net.sayaya.document.data.Model().name(name).cntDocuments(0)))
            .map { data -> MessageModel(MessageModel.MessageType.DELETE, data) }
            .doOnSuccess(publisher::tryEmitNext)
    }
    fun subscribe(): Flux<MessageModel> {
        return subscriber.asFlux()
    }
    private fun map(dto: MessageModel): String {
        return om.writeValueAsString(dto)
    }
    private fun map(json: String): MessageModel {
        return om.readValue(json, MessageModel::class.java)
    }
    @Bean("publish-model")
    fun publishModel(): Supplier<Flux<String>> {
        return Supplier { publisher.asFlux().map(this::map) }
    }
    @Bean("broadcast-model")
    fun broadcastModel(): Consumer<String> {
        return Consumer { json -> subscriber.tryEmitNext(map(json)) }
    }
}